providers/base/bin/smartcard_test.py

```
@@ -0,0 +1,329 @@
+            if self.reader_filter(list_type, r):
+                print(
+                    "smartcard_reader: {}".format(
+                        self.stringify_reader_name(r.name)
+                    )
+                )
+
+    def detect_reader(self, list_type: str):
+        """
+        Detect smart card readers
+
+        :param list_type: contact/contactless/ALL
+
+        """
+        count = 0
+        for r in self.readers:
+            if self.reader_filter(list_type, r):
+                self.logger.info(r.name)
+                count = count + 1
+        if count < 1:
+            raise SystemExit("There is no smartcard reader in this system")
+
+    def get_real_reader_instance(self, reader: str):
+        """
+        Using the stringified smartcard reader name
+        to get the actual reader instance
+
+        :param reader: Stringified smart card reader name
+        """
+        for r in self.readers:
+            if self.stringify_reader_name(r.name) == reader:
+                return r
+
+    def get_connection(self, reader: str):
+        """
+        Connect to the smartcard reader
+
+        :param reader: Stringified smart card reader name
+        """
+        sc_reader = self.get_real_reader_instance(reader)
+        if sc_reader:
+            try:
+                connection = sc_reader.createConnection()
+                connection.connect()
+                self.logger.info("[{}] connected".format(sc_reader))
+                return connection
+            except (NoCardException, CardConnectionException):
+                raise SystemExit("no card inserted or card is unsupported")
+        raise SystemExit("no smartcard reader")
+
+    @timeout(30)
+    def detect_smartcard(self, reader: str):
+        """
+        Detect smartcard insertion and removal in the smartcard reader
+
+        :param reader: Stringified smart card reader name
+        """
+        real_reader = self.get_real_reader_instance(reader)
```

`real_reader` could be None, we can fail early by doing a quick test for None:
```
if real_reader is None:
    raise SystemExit("No real reader was found matching this name: {}".format(reader))
```